[JSC] Introduce non-local wasm inlining decision

Constellation opened the pull request:
#### cfbd6fb2d1c38e82f36f7bd8a8b4cae94bb89a7f
<pre>
[JSC] Introduce non-local wasm inlining decision
<a href="https://bugs.webkit.org/show_bug.cgi?id=301465">https://bugs.webkit.org/show_bug.cgi?id=301465</a>
<a href="https://rdar.apple.com/163406661">rdar://163406661</a>

Reviewed by Dan Hecht.

This patch introduces Wasm::InliningDecision, which overviews the
entire function and deciding which callsite should be inlined. Previous
implementation was deciding inlining locally: we check earlier callsite
and do inlining and we stop inlining when it reaches to the threshold.
Instead, we should check the entire graph of the function and select
inlining candidates globally.

To do this massive design change, we start with V8&apos;s algorithm for
inlining heuristics. We will tune and add more behavior changes after
this but this patch starts with V8&apos;s one. InliningDecision implements
V8&apos;s current inlining heuristics mechanism. And InliningDecision fully
computes which callsite should be inlined a-priori, and OMG compiler
just follows to this decision after all computation is done. We also
increase maxPolymorphicCallees to 4 to align to V8&apos;s heuristics.

* JSTests/wasm/stress/cc-int-to-int-tail-call.js:
* JSTests/wasm/stress/simd-tail-calls-throw.js:
* JSTests/wasm/stress/simple-inline-stacktrace-2.js:
* JSTests/wasm/stress/simple-inline-stacktrace-with-catch-2.js:
* JSTests/wasm/stress/tail-call-across-modules.js:
* JSTests/wasm/stress/tail-call-double.js:
* JSTests/wasm/stress/tail-call-js.js:
* JSTests/wasm/stress/tail-call-simple-i64.js:
* JSTests/wasm/stress/tail-call-simple-int.js:
* JSTests/wasm/stress/tail-call-simple.js:
* JSTests/wasm/stress/tail-call.js:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmCallProfile.h:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::IPIntCallee::IPIntCallee):
(JSC::Wasm::IPIntCallee::needsProfiling const):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmCalleeGroup.cpp:
(JSC::Wasm::CalleeGroup::tryGetBBQCallee):
(JSC::Wasm::CalleeGroup::tryGetBBQCalleeForLoopOSRConcurrently):
* Source/JavaScriptCore/wasm/WasmCalleeGroup.h:
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.h:
(JSC::Wasm::FunctionIPIntMetadataGenerator::addCallTarget):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::addCall):
(JSC::Wasm::IPIntGenerator::addCallIndirect):
(JSC::Wasm::IPIntGenerator::addCallRef):
(JSC::Wasm::IPIntGenerator::finalize):
* Source/JavaScriptCore/wasm/WasmInliningDecision.cpp: Added.
(JSC::Wasm::InliningNode::InliningNode):
(JSC::Wasm::InliningNode::score const):
(JSC::Wasm::InliningDecision::canInline):
(JSC::Wasm::InliningNode::callTarget):
(JSC::Wasm::InliningNode::inlineNode):
(JSC::Wasm::budgetScaleFactor):
(JSC::Wasm::InliningDecision::InliningDecision):
(JSC::Wasm::InliningDecision::profileForCallee):
(JSC::Wasm::isHigherPriority):
(JSC::Wasm::InliningDecision::expand):
* Source/JavaScriptCore/wasm/WasmInliningDecision.h: Added.
(JSC::Wasm::InliningNode::callee const):
(JSC::Wasm::InliningNode::caller const):
(JSC::Wasm::InliningNode::callSites const):
(JSC::Wasm::InliningNode::isInlined const):
(JSC::Wasm::InliningNode::isUnused const):
(JSC::Wasm::InliningNode::caseIndex const):
(JSC::Wasm::InliningNode::depth const):
(JSC::Wasm::InliningNode::callProfileIndex const):
(JSC::Wasm::InliningNode::relativeCallCount const):
(JSC::Wasm::InliningNode::wasmSize const):
* Source/JavaScriptCore/wasm/WasmMergedProfile.cpp:
(JSC::Wasm::MergedProfile::MergedProfile):
(JSC::Wasm::MergedProfile::Candidates::merge):
(JSC::Wasm::MergedProfile::Candidates::finalize const):
(JSC::Wasm::MergedProfile::merge):
* Source/JavaScriptCore/wasm/WasmMergedProfile.h:
(JSC::Wasm::MergedProfile::size const):
(JSC::Wasm::MergedProfile::merged const):
(JSC::Wasm::MergedProfile::totalCount const):
* Source/JavaScriptCore/wasm/WasmModule.cpp:
(JSC::Wasm::Module::createMergedProfile):
* Source/JavaScriptCore/wasm/WasmModule.h:
* Source/JavaScriptCore/wasm/WasmModuleInformation.h:
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::OMGIRGenerator):
(JSC::Wasm::OMGIRGenerator::canInline const):
(JSC::Wasm::OMGIRGenerator::emitInlineDirectCall):
(JSC::Wasm::OMGIRGenerator::emitDirectCall):
(JSC::Wasm::OMGIRGenerator::tryInliningPolymorphicCalls):
(JSC::Wasm::OMGIRGenerator::addCallIndirect):
(JSC::Wasm::OMGIRGenerator::addCallRef):
(JSC::Wasm::parseAndCompileOMG):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
(JSC::Wasm::OMGIRGenerator::OMGIRGenerator):
(JSC::Wasm::OMGIRGenerator::canInline const):
(JSC::Wasm::OMGIRGenerator::addCallIndirect):
(JSC::Wasm::OMGIRGenerator::addCallRef):
* Source/JavaScriptCore/wasm/WasmStreamingParser.cpp:
(JSC::Wasm::StreamingParser::parseFunctionSize):
* Tools/Scripts/run-jsc-stress-tests:

Canonical link: <a href="https://commits.webkit.org/302215@main">https://commits.webkit.org/302215@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/53246ea00e61b29d05c7d2a9bf41ff6c26ab9100

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128358 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/634 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39190 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135750 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79825 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/132943ab-81bf-4528-83f8-c1e6afb77c32) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130230 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/570 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/506 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97707 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/65614 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/54cc7552-cb46-4a34-b5fd-5975d0c45d24) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131306 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/398 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114994 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78300 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6de92b8b-530a-4cae-a2f0-2af428dc81fc) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/376 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33101 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79037 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/120379 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108766 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33585 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138204 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/126818 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/479 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/447 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106241 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/516 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111335 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106045 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/390 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29883 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52796 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20055 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/530 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/63706 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/159842 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/426 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39923 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/492 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/491 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dhecht commented on Source/JavaScriptCore/runtime/OptionsList.h:
```diff
@@ -329,10 +329,14 @@ bool hasCapacityToUseLargeGigacage();
+    v(Unsigned, wasmInliningFactor, 5, Normal, nullptr) \
+    v(Unsigned, wasmInliningBudget, 6000, Normal, nullptr) \
+    v(Unsigned, wasmInliningTinyFunctionThreshold, 12, Normal, nullptr) \
+    v(Unsigned, wasmInliningSmallFunctionThreshold, 50, Normal, nullptr) \
```
consider adding help text for these; it's not obvious what the unit is on all these options.

Constellation replied:
Added.

dhecht commented on Source/JavaScriptCore/wasm/WasmCalleeGroup.cpp:
```diff
@@ -165,6 +165,26 @@ RefPtr<JITCallee> CalleeGroup::tryGetReplacementConcurrently(FunctionCodeIndex f
+    if (m_optimizedCallees.isEmpty())
+        return nullptr;
+
+    // Do not use optimizedCalleesTuple. optimizedCalleesTuple handles currently-installing Callee. But we do not want to handle it actually.
```
Unclear what "handle it" means here. Could you clarify?

Constellation replied:
Changed.

dhecht commented on Source/JavaScriptCore/wasm/WasmInliningDecision.cpp:
```diff
@@ -0,0 +1,276 @@
+
+    // For small-ish functions, the inlining budget is defined by the larger of
+    // 1) the wasm_inlining_min_budget and
+    // 2) the max_growth_factor * initial_wire_byte_size.
```
let's update these to use the correct flag names for easier grep.

Constellation replied:
Changed.

dhecht commented on Source/JavaScriptCore/wasm/WasmInliningDecision.cpp:
```diff
@@ -0,0 +1,276 @@
+    // budget even for large functions that exceed the regular budget.
+    //
+    // Note for future tuning: it might make sense to allow 20% here, and in
+    // turn perhaps lower --wasm-inlining-budget. The drawback is that this
```
ditto

Constellation replied:
Changed.

dhecht commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -6546,23 +6527,36 @@ auto OMGIRGenerator::addCallRef(unsigned callProfileIndex, const TypeDefinition&
-            auto* callee = std::get<0>(candidates.callees()[0]);
-            if (callee->compilationMode() == Wasm::CompilationMode::IPIntMode) {
-                BasicBlock* slowCase = m_proc.addBlock();
+    Vector<ValueResults> fastValuesList;
```
Consider factoring addCallIndirect and addCallRef to share the common code.

Constellation replied:
Done.

dhecht reviewed (approved):
r=me with a few comments

Constellation committed:
[JSC] Introduce non-local wasm inlining decision
https://bugs.webkit.org/show_bug.cgi?id=301465
rdar://163406661

Reviewed by Dan Hecht.

This patch introduces Wasm::InliningDecision, which overviews the
entire function and deciding which callsite should be inlined. Previous
implementation was deciding inlining locally: we check earlier callsite
and do inlining and we stop inlining when it reaches to the threshold.
Instead, we should check the entire graph of the function and select
inlining candidates globally.

To do this massive design change, we start with V8's algorithm for
inlining heuristics. We will tune and add more behavior changes after
this but this patch starts with V8's one. InliningDecision implements
V8's current inlining heuristics mechanism. And InliningDecision fully
computes which callsite should be inlined a-priori, and OMG compiler
just follows to this decision after all computation is done. We also
increase maxPolymorphicCallees to 4 to align to V8's heuristics.

* JSTests/wasm/stress/cc-int-to-int-tail-call.js:
* JSTests/wasm/stress/simd-tail-calls-throw.js:
* JSTests/wasm/stress/simple-inline-stacktrace-2.js:
* JSTests/wasm/stress/simple-inline-stacktrace-with-catch-2.js:
* JSTests/wasm/stress/tail-call-across-modules.js:
* JSTests/wasm/stress/tail-call-double.js:
* JSTests/wasm/stress/tail-call-js.js:
* JSTests/wasm/stress/tail-call-simple-i64.js:
* JSTests/wasm/stress/tail-call-simple-int.js:
* JSTests/wasm/stress/tail-call-simple.js:
* JSTests/wasm/stress/tail-call.js:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmCallProfile.h:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::IPIntCallee::IPIntCallee):
(JSC::Wasm::IPIntCallee::needsProfiling const):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmCalleeGroup.cpp:
(JSC::Wasm::CalleeGroup::tryGetBBQCallee):
(JSC::Wasm::CalleeGroup::tryGetBBQCalleeForLoopOSRConcurrently):
* Source/JavaScriptCore/wasm/WasmCalleeGroup.h:
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.h:
(JSC::Wasm::FunctionIPIntMetadataGenerator::addCallTarget):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::addCall):
(JSC::Wasm::IPIntGenerator::addCallIndirect):
(JSC::Wasm::IPIntGenerator::addCallRef):
(JSC::Wasm::IPIntGenerator::finalize):
* Source/JavaScriptCore/wasm/WasmInliningDecision.cpp: Added.
(JSC::Wasm::InliningNode::InliningNode):
(JSC::Wasm::InliningNode::score const):
(JSC::Wasm::InliningDecision::canInline):
(JSC::Wasm::InliningNode::callTarget):
(JSC::Wasm::InliningNode::inlineNode):
(JSC::Wasm::budgetScaleFactor):
(JSC::Wasm::InliningDecision::InliningDecision):
(JSC::Wasm::InliningDecision::profileForCallee):
(JSC::Wasm::isHigherPriority):
(JSC::Wasm::InliningDecision::expand):
* Source/JavaScriptCore/wasm/WasmInliningDecision.h: Added.
(JSC::Wasm::InliningNode::callee const):
(JSC::Wasm::InliningNode::caller const):
(JSC::Wasm::InliningNode::callSites const):
(JSC::Wasm::InliningNode::isInlined const):
(JSC::Wasm::InliningNode::isUnused const):
(JSC::Wasm::InliningNode::caseIndex const):
(JSC::Wasm::InliningNode::depth const):
(JSC::Wasm::InliningNode::callProfileIndex const):
(JSC::Wasm::InliningNode::relativeCallCount const):
(JSC::Wasm::InliningNode::wasmSize const):
* Source/JavaScriptCore/wasm/WasmMergedProfile.cpp:
(JSC::Wasm::MergedProfile::MergedProfile):
(JSC::Wasm::MergedProfile::Candidates::merge):
(JSC::Wasm::MergedProfile::Candidates::finalize const):
(JSC::Wasm::MergedProfile::merge):
* Source/JavaScriptCore/wasm/WasmMergedProfile.h:
(JSC::Wasm::MergedProfile::size const):
(JSC::Wasm::MergedProfile::merged const):
(JSC::Wasm::MergedProfile::totalCount const):
* Source/JavaScriptCore/wasm/WasmModule.cpp:
(JSC::Wasm::Module::createMergedProfile):
* Source/JavaScriptCore/wasm/WasmModule.h:
* Source/JavaScriptCore/wasm/WasmModuleInformation.h:
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::OMGIRGenerator):
(JSC::Wasm::OMGIRGenerator::canInline const):
(JSC::Wasm::OMGIRGenerator::emitInlineDirectCall):
(JSC::Wasm::OMGIRGenerator::emitDirectCall):
(JSC::Wasm::OMGIRGenerator::tryInliningPolymorphicCalls):
(JSC::Wasm::OMGIRGenerator::addCallIndirect):
(JSC::Wasm::OMGIRGenerator::addCallRef):
(JSC::Wasm::parseAndCompileOMG):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
(JSC::Wasm::OMGIRGenerator::OMGIRGenerator):
(JSC::Wasm::OMGIRGenerator::canInline const):
(JSC::Wasm::OMGIRGenerator::addCallIndirect):
(JSC::Wasm::OMGIRGenerator::addCallRef):
* Source/JavaScriptCore/wasm/WasmStreamingParser.cpp:
(JSC::Wasm::StreamingParser::parseFunctionSize):
* Tools/Scripts/run-jsc-stress-tests:

Canonical link: https://commits.webkit.org/302215@main
